Applications/FileIO/SWMM/SWMMInterface.h

```diff
@@ -0,0 +1,245 @@
+    class Element;
+}
+
+/// SWMM object types
+enum class SwmmObject
+{
+    SUBCATCHMENT = 0,
+    NODE = 1,
+    LINK = 2,
+    SYSTEM = 3
+};
+
+/**
+ * Interface for reading files used within the  * Storm Water Management Model (SWMM) and
+ * converting the data therein into corresponding OGS data structures.
+ * SWMM distinguishes four different object types (defined in the SwmmObject enum):
+ *     * Subcatchments
+ *     * Nodes (or junctions)
+ *     * Links (or conduits)
+ *     * System
+ * Note that each object in a SWMM input file has a name.
+ * Each of the four object types has a (different) number of parameters. There can be an arbitrary
+ * number of nodes, links or subcatchments but only ever one system.
+ * The interface can convert the SWMM input data into a geometry or a (line-)mesh.
+ * For meshes, also output data can be added to nodes and elements.
+ * The interface also provides routines for returning data as vectors as well as convenience
+ * methods for outputting data into CSV files. CSV files will either contain all parameters for one
+ * object at all time steps or all parameters for all objects of a given type at one timestep.
+ */
+class SwmmInterface
+{
+public:
+    /// Basic method to create the interface (containing a mesh) from a SWMM file
+    /// Other non-static methods rely on this being called in the beginning.
+    static SwmmInterface* create(std::string const& file_name);
+
+    /// If a mesh has already been created, this methods allows to add node- or link-arrays as property
+    /// to that mesh. The data vectors can be created using the getArrayAtTimeStep() method.
+    static bool addResultsToMesh(MeshLib::Mesh &mesh, SwmmObject const type,
+        std::string const& vec_name, std::vector<double> const& data);
+
+    /// Returns the mesh generated from SWMM file content.
+    MeshLib::Mesh* getMesh() const { return _mesh.get(); }
+
+    /// Returns the name of the data array for the given object type and parameter index.
+    std::string getArrayName(SwmmObject obj_type, std::size_t var_idx) const;
+
+    /// Get all the object names for a given object type
+    std::vector<std::string> getNames(SwmmObject obj_type) const;
+
+    /// Returns the Name for the indexed object of the given type (or an empty string if an error occured).
+    std::string getName(SwmmObject obj_type, std::size_t idx) const;
+
+    /// Returns the number of objects of the given type.
+    std::size_t getNObjects(SwmmObject obj_type) const;
+
+    /// Returns the number of parameters (incl. pollutants) of the given type.
+    std::size_t getNParameters(SwmmObject obj_type) const;
+
+    /// Returns the number of time steps for the simulation results.
+    std::size_t getNTimeSteps() const;
+
+    /// Returns an array for a given variable at all nodes/links from a SWMM output file for a given time step.
+    std::vector<double> getArrayAtTimeStep(SwmmObject obj_type, std::size_t time_step, std::size_t var_idx) const;
```

Comment:
Yes, I have to. The vectors do not exist as such. So if the user asks for a certain vector, it is put together on the fly and a copy is returned. If I returned a reference, it'd be instantly gone.